[quant][graphmode][fix] Insert dequantize before use node

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35130 [quant][graphmode] Make interpolate/upsample work again
* #35142 [quant][graphmode] SwapDeQuant support prim::If
* #35141 [quant][graphmode][refactor] getGeneralOpTensorInputIndexes -> getGeneralOpTensorInputs
* #35135 [quant][graphmode][refactor] swapDeQuant takes block as arugment
* #35077 [quant][graphmode] Fold quantized prepacking ops
* #35073 [quant] Make conv2d_prepack and linear_prepack pure
* #34855 [quant][graphmode] quantization support for prim::CallFunction
* #34346 [quant][graphmode] Add quantization support for aten::cat
* #34345 [quant][graphmode] Add prim::ListConstruct to general op handling
* #34854 [quant][graphmode] quantization support for view, transpose, contiguos
* #34806 [quant][graphmode] quantization support for aten::chunk
* #34807 [quant][graphmode] quantization support for prim::ListUnpack
* #34804 [quant][graphmode] Replicate quantize node for prim::If
* #34572 [quant][graphmode] quantization support for aten::add
* #34518 [quant][graphmode] quantization work for prim::If
* #34571 [quant][graphmode][fix] use observed_values_ to check values are observed
* #34414 [quant][graphmode][refactor] insertObservers for Block
* **#34411 [quant][graphmode][fix] Insert dequantize before use node**
* #34349 [quant][graphmode][fix] preserve the type of original value when inserting dequant node

Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D20519603](https://our.internmc.facebook.com/intern/diff/D20519603)

jerryzh168 committed:
[quant][graphmode][fix] Insert dequantize before use node

Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on test/test_jit.py:
```diff
@@ -1865,6 +1865,43 @@ def forward(self, x):
+        # check dequantize is right before add
+        FileCheck().check("aten::dequantize") \
+                   .check("aten::dequantize") \
+                   .check_next("aten::add") \
```
How does the graph have add? From what I can see above, you are setting `cond` to True so it should have a conv.

jerryzh168 replied:
both blocks of prim::If will be quantized

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

z-a-f reviewed (approved):
LGTM

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D20519603](https://our.internmc.facebook.com/intern/diff/D20519603)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D20519603](https://our.internmc.facebook.com/intern/diff/D20519603)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D20519603](https://our.internmc.facebook.com/intern/diff/D20519603)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D20519603](https://our.internmc.facebook.com/intern/diff/D20519603)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D20519603](https://our.internmc.facebook.com/intern/diff/D20519603)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D20519603](https://our.internmc.facebook.com/intern/diff/D20519603)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D20519603](https://our.internmc.facebook.com/intern/diff/D20519603)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D20519603](https://our.internmc.facebook.com/intern/diff/D20519603)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D20519603](https://our.internmc.facebook.com/intern/diff/D20519603)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D20519603](https://our.internmc.facebook.com/intern/diff/D20519603)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D20519603](https://our.internmc.facebook.com/intern/diff/D20519603)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D20519603](https://our.internmc.facebook.com/intern/diff/D20519603)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D20519603](https://our.internmc.facebook.com/intern/diff/D20519603)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D20519603](https://our.internmc.facebook.com/intern/diff/D20519603)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Insert dequantize before use node"


Summary:
To make sure dequantize and the node that uses the dequantized value reside in the
block, so that we can do quant fusion

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D20519603](https://our.internmc.facebook.com/intern/diff/D20519603)

[ghstack-poisoned]
